sdk/core/core/src/az_json_builder.c

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <az_json_builder.h>
+
+#include <az_str.h>
+
+#include <_az_cfg.h>
+
+AZ_NODISCARD az_result
+az_json_builder_init(az_json_builder * const out, az_span_action const write) {
+  AZ_CONTRACT_ARG_NOT_NULL(out);
+
+  *out = (az_json_builder){ .write = write };
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD static az_result
+az_json_builder_write_str(az_json_builder * const self, az_span const value) {
+  AZ_CONTRACT_ARG_NOT_NULL(self);
+
+  az_span_action const write = self->write;
+  AZ_RETURN_IF_FAILED(az_span_action_do(write, AZ_STR("\"")));
+  AZ_RETURN_IF_FAILED(az_span_action_do(write, value));
+  AZ_RETURN_IF_FAILED(az_span_action_do(write, AZ_STR("\"")));
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result
+az_json_builder_write(az_json_builder * const self, az_json_value const value) {
+  AZ_CONTRACT_ARG_NOT_NULL(self);
+
+  az_span_action const write = self->write;
+  switch (value.kind) {
+    case AZ_JSON_VALUE_NULL: {
+      self->need_comma = true;
+      return az_span_action_do(write, AZ_STR("null"));
+    }
+    case AZ_JSON_VALUE_BOOLEAN: {
+      self->need_comma = true;
+      return az_span_action_do(write, value.data.boolean ? AZ_STR("true") : AZ_STR("false"));
+    }
+    case AZ_JSON_VALUE_STRING: {
+      self->need_comma = true;
+      return az_json_builder_write_str(self, value.data.string);
```

Comment:
I think you want this to be your "master" function. That is, this is the function that writes to the json_builder. And, then you have convenience methods that call this method. So, put in this case the code to write the string to the json_builder and have the convenient az_json_builder_write_str function call this function.